server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java

```diff
@@ -22,12 +22,114 @@
 import org.keycloak.models.RequiredActionProviderModel;
 import org.keycloak.models.UserModel;
 
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static org.keycloak.common.Profile.isFeatureEnabled;
+
 /**
  * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
  * @version $Revision: 1 $
  */
 public class DefaultRequiredActions {
+
+    /**
+     * Disable required actions, which should not be present due to environment dependent reasons (f.e. feature is disabled)
+     *
+     * @param realm realm
+     */
+    public static void manageEnvironmentDependentActions(RealmModel realm) {
+        realm.getRequiredActionProvidersStream().forEach(f -> manageEnvironmentDependentAction(realm, f));
+    }
+
+    /**
+     * Disable required actions, which should not be present due to environment dependent reasons (f.e. feature is disabled)
+     *
+     * @param realm    realm
+     * @param provider required action provider
+     */
+    public static void manageEnvironmentDependentAction(RealmModel realm, RequiredActionProviderModel provider) {
+        if (provider == null) return;
+
+        final Optional<Action> foundAction = Action.findByAlias(provider.getAlias());
+        foundAction.ifPresent((action) -> {
+            if (action.shouldBeExplicitlyDisabled() && provider.isEnabled()) {
+                provider.setEnabled(false);
```

Comment:
I am not sure if it is good to explicitly update DB this way.

For example when there is this scenario:
1) WebAuthn feature is enabled and WebAuthn required actions are enabled in the realm
2) Now administrator restarts Keycloak with WebAuthn feature disabled
3) Some user tries to authenticate, which will explicitly disable WebAuthn actions in the DB
4) Now administrator restarts Keycloak again with WebAuthn feature enabled. But now the WebAuthn required actions are disabled in the DB (which administrator may not expect). Also it may happen that WebAuthn actions were auto-disabled just for some realms (EG. If point 3 did not happen for the realm `foo` and none of the user tried to authenticate to this realm when WebAuthn feature was disabled, then the requiredAction will be still enabled for this realm. Which means a bit of mess IMO as some realms may have required actions still enabled and some were automatically disabled etc...)

IMO if we go with this approach, it will be better that during authentication, the required action with non-existent provider factory will be ignored and there won't be any DB update at all. In this case, it will be also good to add the WARNING into the server.log that realm contains some required action with disabled (non-existent factory).

Reply:
@mposolda Agree with that! It's a better approach. Thanks for your review! :))